core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java

```diff
@@ -38,207 +40,221 @@
    */
-  public DefaultBlockWorkerClient(UserState userState, GrpcServerAddress address,
-      AlluxioConfiguration alluxioConf) throws IOException {
-    RetryPolicy retryPolicy = RetryUtils.defaultClientRetry();
-    UnauthenticatedException lastException = null;
-    // TODO(feng): unify worker client with AbstractClient
-    while (retryPolicy.attempt()) {
-      try {
-        // Disables channel pooling for data streaming to achieve better throughput.
-        // Channel is still reused due to client pooling.
-        mStreamingChannel = GrpcChannelBuilder.newBuilder(address, alluxioConf)
-            .setSubject(userState.getSubject())
+  public DefaultBlockWorkerClient(ClientContext context, GrpcServerAddress address,
+                                  AlluxioConfiguration alluxioConf) {
+    // BlockWorkerClient is typically used inside AlluxioFileInStream
+    // to fetch block chunk. AlluxioFilInStream has its own retry mechanism, so we don't
+    // retry at this level so as not to interfere with the retry of our higher-level
+    // callers.
+    // An instance where retries interfere with each other is when we and our calling
+    // AlluxioFileInStream both adopted a timer-based retry, then all the time could be wasted
+    // in our inner retry, and the outer stream has no chance of retrying a different client.
+    super(context, () -> new CountingRetry(0));
+    // the server address this client connects to. It might not be an IP address
+    mAddress = address;
+    mRpcTimeoutMs = alluxioConf.getMs(PropertyKey.USER_RPC_RETRY_MAX_DURATION);
+  }
+
+  @Override
+  protected void beforeConnect() {}
+
+  @Override
+  protected void afterConnect() {
+    // set up stubs
+    mStreamingAsyncStub = BlockWorkerGrpc.newStub(mStreamingChannel);
+    mRpcFutureStub = BlockWorkerGrpc.newFutureStub(mChannel);
+    mRpcBlockingStub = BlockWorkerGrpc.newBlockingStub(mChannel);
+  }
+
+  @Override
+  protected void doConnect() throws AlluxioStatusException {
+    super.doConnect();
+
+    // build a streaming channel
+    AlluxioConfiguration conf = mContext.getClusterConf();
+    mStreamingChannel = GrpcChannelBuilder.newBuilder(mServerAddress, conf)
+            .setSubject(mContext.getSubject())
             .setNetworkGroup(GrpcNetworkGroup.STREAMING)
             .build();
-        mStreamingChannel.intercept(new StreamSerializationClientInterceptor());
-        // Uses default pooling strategy for RPC calls for better scalability.
-        mRpcChannel = GrpcChannelBuilder.newBuilder(address, alluxioConf)
-            .setSubject(userState.getSubject())
-            .setNetworkGroup(GrpcNetworkGroup.RPC)
-            .build();
-        lastException = null;
-        break;
-      } catch (StatusRuntimeException e) {
-        close();
-        throw AlluxioStatusException.fromStatusRuntimeException(e);
-      } catch (UnauthenticatedException e) {
-        close();
-        userState.relogin();
```

Comment:
How about retrying on `UnauthenticatedException` unconditionally? It won't work to distinguish between initial connection and other connections, it would still break if the worker failed before the first connection is made.